docs(world): add the systems page to new docs. (#1836)
latticexyz/mud: next-docs/pages/world/systems.mdx

# Systems

> One of the design principles of MUD is to separate the state of the `World` from the business logic that it implements.

Reviewer: ```suggestion
One of the design principles of MUD is to separate the state of the `World` from the business logic.
```

> We do this by having stateless `System` contracts with the functions that implement the business logic.
> These contracts are called through the `World`, and call back on the `World` when they need to read or modify state information, which is stored in [tables](./tables).

Reviewer: ```suggestion
The business logic is implemented in stateless `System` contracts.
`System`s are called through the `World`, and call back to the `World` to read and write state from [tables](./tables).
```

<details>

<summary>Detailed illustration</summary>

![Detailed illustration](./world-table.svg)

> 1. An external entity (contract or externally owned account) calls on the `World` a function called `myns_sys_fun`.

Reviewer: We try to avoid abbreviations everywhere, should also do it in the docs
```suggestion
1. An external entity (contract or externally owned account) calls on the `World` a function called `namespace_system_function`.
```

>    This means the `fun` function in the `sys` system, which is part of the `myns` namespace (the system is also refered to as `myns:sys`).

Reviewer: ```suggestion
   [This function was registered](./function-selectors) by the owner of the `namespace` namespace and points to the `function` function in the `system` system in the `namespace` namespace.
```

1. The `World` verifies that access is permitted (for example, because `myns:sys` is publicly accessible) and if so calls `fun` on the `myns:sys` contract with the provided parameters.

1. At some point in its execution `fun` decides to update the data in the table `myns:tab`.
   As with all other tables, this table is stored in the `World`'s storage.
   To modify it, `fun` calls a function on the `World` contract.

> 1. The `World` verifies that access is permitted (by default it would be, because `myns:sys` has access to the `myns` namespace).

Reviewer: ```suggestion
1. The `World` verifies that access is permitted (in this example it would be, because [systems have access to their namespace](./namespace-access-control)).
```

   If so, it modifies the data in the `myns:tab` table.

</details>

The `World` serves as a central entry point and forwards calls to systems, which allows it to provide [access control](/world/namespaces-access-control).

## Calling systems

To call a `System`, you call the `World` in one of these ways:

> - If a function is [registered](https://github.com/latticexyz/mud/blob/main/packages/world/src/modules/core/implementations/WorldRegistrationSystem.sol#L164-L201), you can call it using the syntax `world.namespace_system_function(parameters)`.

Reviewer: ```suggestion
- If a [function selector for the `System` is registered in the World](./function-selectors), you can call it via `world.<namespace>_<system>_<function>(<arguments>)`.
```

- You can use [`call`](https://github.com/latticexyz/mud/blob/main/packages/world/src/World.sol#L346-L351).
- If you have [the proper delegation](/world/account-delegation) you can use [`callFrom`](https://github.com/latticexyz/mud/blob/main/packages/world/src/World.sol#L353-L394).

## Writing systems

> A `System` should _not_ have any state internally, but store all of it in tables in the `World`.

Reviewer: ```suggestion
A `System` should _not_ have any internal state, but store all of it in tables in the `World`.
```

There are several reasons for this:

- It allows a `World` to enforce access controls.
- It allows the same `System` to be used by multiple `World` contracts.
- Upgrades are a lot simpler when all the state is centralized outside of the `System` contract.

> Because calls to systems are proxied by the `World`, some message fields don't reflect the original call.

Reviewer: ```suggestion
Because calls to systems are proxied through the `World`, some message fields don't reflect the original call.
```

Use these substitutes:

> | Normal code  | `System` replacement |

Reviewer: ```suggestion
| Vanilla Solidity  | `System` replacement |
```

| ------------ | -------------------- |
| `msg.sender` | `_msgSender()`       |
| `msg.value`  | `_msgValue()`        |

> When calling other contracts from a `System`, be aware that if you use `delegatecall` the called contract inherits the caller's permissions and can modify data from the `World` on behalf of the `System`.

Reviewer: ```suggestion
When calling other contracts from a `System`, be aware that if you use `delegatecall` the called contract inherits the `System`'s permissions and can modify data in the `World` on behalf of the `System`.
```

| `_msgValue()`                                | zero                                                                                                                                            | can use [`WorldContextProvider`](https://github.com/latticexyz/mud/blob/main/packages/world/src/WorldContext.sol#L180-L203) to transfer the correct information |
| Can be used by systems in the root namespace | No (it's a security measure)                                                                                                                    | Yes                                                                                                                                                             |

## Registering systems

For a `System` to be callable from a `World` it has to be [registered](https://github.com/latticexyz/mud/blob/main/packages/world/src/modules/core/implementations/WorldRegistrationSystem.sol#L97-L201).

> If the namespace in which the `System` is registered already exists, only the [namespace owner](/world/namespaces-access-control#ownership).
> If it doesn't exist yet it is created, and the address that called `registerSystem` is assigned as the owner of the new namespace.

Reviewer: Note that we're going to remove the auto-creation of namespaces shortly after the audit because it leads to frontrunning issues. My preference would be to already not mention it here.

```suggestion
Only the [namespace owner](./namespaces-access-control#ownership) can register a `System` in a namespace.
```

> In a `World` there can only be one `System` per contract address.
> If you need the functionality and there is already a `System` registered for that address, you can always deploy the bytecode as a new contract.

Reviewer: ```suggestion
`System`s can be registered once per `World`, but the same system can be registered in multiple `World`s.
If you need multiple instances of a `System` in the same world, you can deploy the `System` multiple times and register the individual deployments individually.
```

### Upgrading systems

The namespace owner can upgrade a `System`.
This is a two-step process: deploy the contract for the new `System` and then call `registerSystem` with the same `ResourceId` as the old one and the new contract address.

This upgrade process removes the old `System` contract's access to the namespace, and gives access to the new contract.
Any access granted _manually_ to the `System` is not revoked, nor granted to the upgraded `System`.

> **Note:** You _should_ make sure to remove any such access.

Reviewer: ```suggestion
**Note:** You _should_ make sure to remove any such manually granted access.
```

`System` access is based on the contract address, so somebody else could register a namespace they'd own, register the old `System` contract as a system in their namespace, and then abuse those permissions (if the `System` has code that can be used for that, of course).

## Access control

When you register a `System`, you can specify whether it is going to be private or public.

- A public `System` has no access control checks, it can be called by anybody.
  This is the main mechanism for user interaction with a MUD application.

> - A private `System` can only be called by entities that have access.

Reviewer: ```suggestion
- A private `System` can only be called by accounts that have access.
```

  This access can be the result of:

  - Access permission to the namespace in which the `System` is registered.
  - Access permission specifically to the `System`.

>   Note that while code in private systems can only be called by authorized resources, by default those authorized resources include public systems in the same namespace operating on behalf of any user.

Reviewer: ```suggestion
Note that `System`s have access to their own namespace by default, so public `System`s can call private `System`s in their namespace.
```

>   Vulnerabilities in private systems are _harder_ to exploit, but not necessarily impossible.

Reviewer: ```suggestion
```

## Root systems

The `World` uses `call` for systems in other namespaces, but `delegatecall` for those in the root namespace (`bytes14(0)`).
As a result, root systems have access to the `World` contract's storage.
Because of this access, root systems use [the internal `StoreCore` methods](https://github.com/latticexyz/mud/blob/main/packages/store/src/StoreCore.sol), which are slightly cheaper than calling [the `IStore` methods](https://github.com/latticexyz/mud/blob/main/packages/store/src/IStore.sol) used by other systems.
Note that the table libraries abstract this difference, so normally there is no reason to be concerned about it.

Another effect of having access to the storage of the `World` is that root systems could, in theory, overwrite any information in any table regardless of access control.

> We recommend that the root namespace be used only when needed, and any `System` registered for it be properly audited.
> Of course, only the owner of the root namespace can register systems in the root namespace.

Reviewer: Slight rephrasing. (Removed the part about auditing systems, sounded like we recommend _only_ root systems to be audited, but ideally all systems should be audited)
```suggestion
Only the [owner of the root namespace](./namespaces-access-control) can register root systems.
We recommend to only use the root namespace when strictly necessary.
```
